In [52]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as pyplot
import seaborn as sns

In [53]:
file_path = "C:\\Users\\Kishan D K\\OneDrive\\Documents\\Data_science\\Data_files\\visadataset.csv"
file_path

'C:\\Users\\Kishan D K\\OneDrive\\Documents\\Data_science\\Data_files\\visadataset.csv'

In [54]:
visa_df = pd.read_csv(file_path)
visa_df

case_id continent education_of_employee has_job_experience  \
0    EZYV00001    Europe           High School                  N   
1    EZYV00002  North Am            Bachelor's                  N   
2    EZYV00003      Asia           High School                  Y   
3    EZYV00004  North Am            Bachelor's                  N   
4    EZYV00005    Africa             Doctorate                  N   
..         ...       ...                   ...                ...   
195  EZYV00196    Africa            Bachelor's                  Y   
196  EZYV00197    Europe              Master's                  Y   
197  EZYV00198    Africa           High School                  Y   
198  EZYV00199  North Am              Master's                  N   
199  EZYV00200    Africa            Bachelor's                  Y   

    requires_job_training  no_of_employees  yr_of_estab region_of_employment  \
0                       Y            12815         1986                South   
1                       Y            58485         1955            Northeast   
2                       Y            31262         1904              Midwest   
3                       Y            42617         1993              Midwest   
4                       Y            69628         1909            Northeast   
..                    ...              ...          ...                  ...   
195                     N            14629         2004                 West   
196                     Y             7504         2012                 West   
197                     Y            43149         1900                South   
198                     Y           111098         1888                South   
199                     N           140121         1981              Midwest   

     prevailing_wage unit_of_wage full_time_position case_status  
0          100248.93         Year                  N   Certified  
1           36080.79         Year                  Y   Certified  
2          117927.92         Year                  N   Certified  
3          110256.05         Hour                  N   Certified  
4          135456.27         Hour                  Y      Denied  
..               ...          ...                ...         ...  
195         91543.97         Hour                  Y      Denied  
196         80136.89         Hour                  Y      Denied  
197        100729.22         Year                  N   Certified  
198         74387.16         Year                  Y      Denied  
199         46578.32         Year                  Y      Denied  

[200 rows x 12 columns]

In [55]:
visa_df.columns

Index(['case_id', 'continent', 'education_of_employee', 'has_job_experience',
       'requires_job_training', 'no_of_employees', 'yr_of_estab',
       'region_of_employment', 'prevailing_wage', 'unit_of_wage',
       'full_time_position', 'case_status'],
      dtype='object')

In [56]:
visa_df['prevailing_wage']   # Series

0      100248.93
1       36080.79
2      117927.92
3      110256.05
4      135456.27
         ...    
195     91543.97
196     80136.89
197    100729.22
198     74387.16
199     46578.32
Name: prevailing_wage, Length: 200, dtype: float64

In [57]:
visa_df['prevailing_wage'].values

array([100248.93,  36080.79, 117927.92, 110256.05, 135456.27, 145905.88,
        55120.87,  71287.67, 129869.8 , 111859.19, 124064.18,  93976.01,
       132684.56,  55114.74,  32508.72,  77223.39,  77344.56, 116006.58,
        30070.73,  69373.6 , 113577.54,  87655.47,  90587.64, 144839.63,
       108551.73, 131019.3 , 145128.76, 148423.43,  33490.9 ,  66311.92,
       120208.59,  74216.68,  41522.71, 127039.5 , 124525.97, 144955.85,
        39081.98, 122972.66,  80512.47, 107809.41,  71706.76, 125798.54,
        85057.22,  73069.74,  51650.7 ,  64542.62,  50046.51, 114201.48,
       140560.73,  72409.86,  40709.44, 147768.9 , 104931.92, 110057.44,
        71583.76, 141889.69,  91687.89, 109554.31, 136617.36, 141249.86,
       115249.65,  85725.07, 141697.55,  66065.04,  97660.5 ,  98739.87,
        89596.78,  68906.02,  88214.65,  60950.32, 147143.31, 126785.22,
        38089.88, 144385.97, 107345.82,  50569.92, 130022.43,  68918.29,
        56992.46, 120652.75, 142096.36, 124027.27, 

In [58]:
visa_df.dtypes

case_id                   object
continent                 object
education_of_employee     object
has_job_experience        object
requires_job_training     object
no_of_employees            int64
yr_of_estab                int64
region_of_employment      object
prevailing_wage          float64
unit_of_wage              object
full_time_position        object
case_status               object
dtype: object

- count
- min 
- max
- mean
- median


In [59]:
wage = visa_df['prevailing_wage']
wage.count()

200

In [60]:
wage.min()

30070.73

In [61]:
wage.max()

148423.43

In [62]:
wage.mean()

91937.0197

In [63]:
wage.median()

91678.54000000001

In [64]:
dict1 = {}                                                    #creating empty dictionary
wage_count = round(visa_df['prevailing_wage'].count(),2)      # Counts how many values are in the column prevailing_wage and rounds it to 2 decimals.
wage_min = round(visa_df['prevailing_wage'].min(),2)          #Finds the minimum wage value and rounds it.
wage_max = round(visa_df['prevailing_wage'].max(),2)
wage_mean = round(visa_df['prevailing_wage'].mean(),2)                                                     #Finds the maximum wage value and rounds it.
wage_median = round(visa_df['prevailing_wage'].median(),2)    #Finds the median (middle) wage value and rounds it.
wage_std = round(visa_df['prevailing_wage'].std(),2)
dict1['count']=wage_count
dict1['min']=wage_min                                         #Stores all the calculated values into the dictionary with proper keys.
dict1['max']=wage_max
dict1['mean']=wage_mean
dict1['median']=wage_median
dict1['std']=wage_std
dict1                                                         #Displays the dictionary with results.
pd.DataFrame(dict1,index=['prevailing_wage'])                 #Converts the dictionary into a DataFrame (table) with one row named "prevailing_wage".

count       min        max      mean    median       std
prevailing_wage    200  30070.73  148423.43  91937.02  91678.54  33715.17

In [65]:
wage_count = round(visa_df['prevailing_wage'].count(),2)      
wage_min = round(visa_df['prevailing_wage'].min(),2)          
wage_max = round(visa_df['prevailing_wage'].max(),2)
wage_mean = round(visa_df['prevailing_wage'].mean(),2)   
wage_median = round(visa_df['prevailing_wage'].median(),2)    
wage_std = round(visa_df['prevailing_wage'].std(),2)

list1 = [wage_count,wage_min,wage_max,wage_mean,wage_median,wage_std]

pd.DataFrame(list1,
             columns = ['prevailing_wage'],
             index = ['count','min','max','mean','median','std'])

prevailing_wage
count            200.00
min            30070.73
max           148423.43
mean           91937.02
median         91678.54
std            33715.17

In [66]:
#step-1: numerical column list
dtypes=dict(visa_df.dtypes)
num=[i for i in dtypes if dtypes[i]!='O']
print(num)

dict1={}
for i in num:
    count=round(visa_df[i].count(),2)
    MIN=round(visa_df[i].min(),2)
    MAX=round(visa_df[i].max(),2)
    mean=round(visa_df[i].mean(),2)
    median=round(visa_df[i].median(),2)
    std=round(visa_df[i].std(),2)
    list1=[count,MIN,MAX,mean,median,std]
    dict1[i]=list1

df=pd.DataFrame(dict1,
             index=['count','min','max','mean','median','std'])
dict1

['no_of_employees', 'yr_of_estab', 'prevailing_wage']


{'no_of_employees': [200, 693, 147663, 77312.69, 75295.5, 41787.3],
 'yr_of_estab': [200, 1880, 2022, 1949.67, 1951.5, 40.93],
 'prevailing_wage': [200, 30070.73, 148423.43, 91937.02, 91678.54, 33715.17]}

In [67]:
visa_df.describe()

no_of_employees  yr_of_estab  prevailing_wage
count       200.000000   200.000000        200.00000
mean      77312.690000  1949.670000      91937.01970
std       41787.295918    40.933223      33715.17152
min         693.000000  1880.000000      30070.73000
25%       43128.500000  1912.000000      64444.56750
50%       75295.500000  1951.500000      91678.54000
75%      112360.250000  1986.000000     120905.87750
max      147663.000000  2022.000000     148423.43000

In [68]:
visa_df['prevailing_wage'].mean()      #reading a specific column and applying methods....like mean, mode, median...etc

91937.0197

In [69]:
visa_df['prevailing_wage'].median()

91678.54000000001

In [70]:
visa_df['prevailing_wage'].mode()

0       30070.73
1       32508.72
2       32676.59
3       32892.47
4       33490.90
         ...    
195    147143.31
196    147155.12
197    147768.90
198    148068.00
199    148423.43
Name: prevailing_wage, Length: 200, dtype: float64

*****percentile and quantile*****

- percentile range from 1 to 100
- quantile q1 = 25p, q2 = 50p, q3 = 75p
- np.percentile(<direct number between 1 to 100>, data)
- example = np.percentile(75,data)
- np.quantile(<decimal number>, data)
- example = np.quantile(0.75,data)

In [71]:
q1 = round(np.percentile(visa_df['prevailing_wage'],25),2)
q2 = round(np.percentile(visa_df['prevailing_wage'],50),2)
q3 = round(np.percentile(visa_df['prevailing_wage'],75),2)
print(q1,q2,q3)

64444.57 91678.54 120905.88


In [72]:
q1 = round(np.quantile(visa_df['prevailing_wage'],0.25),2)
q2 = round(np.quantile(visa_df['prevailing_wage'],0.50),2)
q3 = round(np.quantile(visa_df['prevailing_wage'],0.75),2)
print(q1,q2,q3)

64444.57 91678.54 120905.88


In [73]:
dtypes=dict(visa_df.dtypes)
num=[i for i in dtypes if dtypes[i]!='O']
print(num)

dict1={}
for i in num:
    count=round(visa_df[i].count(),2)
    MIN=round(visa_df[i].min(),2)
    MAX=round(visa_df[i].max(),2)
    mean=round(visa_df[i].mean(),2)
    median=round(visa_df[i].median(),2)
    std=round(visa_df[i].std(),2)
    ###########################################################
    q1 = round(np.percentile(visa_df[i],0.25),2)
    q2 = round(np.percentile(visa_df[i],0.50),2)
    q3 = round(np.percentile(visa_df[i],0.75),2)

    
    list1=[count,MIN,MAX,mean,median,std,q1,q2,q3]
    dict1[i]=list1

    df=pd.DataFrame(dict1,
                 index=['count','min','max','mean','median','std','25','50','75'])
dict1

['no_of_employees', 'yr_of_estab', 'prevailing_wage']


{'no_of_employees': [200,
  693,
  147663,
  77312.69,
  75295.5,
  41787.3,
  806.93,
  920.86,
  1229.32],
 'yr_of_estab': [200,
  1880,
  2022,
  1949.67,
  1951.5,
  40.93,
  1880.0,
  1880.0,
  1880.49],
 'prevailing_wage': [200,
  30070.73,
  148423.43,
  91937.02,
  91678.54,
  33715.17,
  31283.63,
  32496.53,
  32591.4]}

df

In [74]:
q1 = round(np.percentile(visa_df['prevailing_wage'],25),2)
q1

64444.57

In [76]:
25*156192.00/100

39048.0

******here what is the meaning of that value 64444.57 is ?******

******25% of data below 64444.57******

******25% of the observations from the total data have a value below 64444.57******

In [79]:
visa_df['prevailing_wage'].count()

200

In [81]:
25*200/100

50.0

In [84]:
len([visa_df['prevailing_wage']<64444.57])

1

$Empiricle$-$rule$

<img src="https://builtin.com/sites/www.builtin.com/files/styles/ckeditor_optimize/public/inline-images/1_empirical-rule.jpg" jsaction="" class="sFlh5c FyHeAf iPVvYb" style="max-width: 810px; height: 152px; margin: 0px; width: 304px;" alt="Empirical Rule (68-95-99.7) Explained | Built In" jsname="kn3ccd">

**when data shows a normal distribution**
- u-1*sigma to u+1*sigma : 68%
- u-2*sigma to u+2*sigma : 95%
- u-3*sigma to u+3*sigma : 99.7%

In [89]:
wage_mean,  wage_std

(91937.02, 33715.17)

In [91]:
# wage_mean = u
# wage_std = sigma

In [93]:
val_minus_1 = round(wage_mean-1*wage_std,2)
val_plus_1 = round(wage_mean+1*wage_std,2)
val_minus_2 = round(wage_mean-2*wage_std,2)
val_plus_2 = round(wage_mean+2*wage_std,2)
val_minus_3 = round(wage_mean-3*wage_std,2)
val_plus_3 = round(wage_mean+3*wage_std,2)

print(val_minus_1,val_plus_1,val_minus_2,val_plus_2,val_minus_3,val_plus_3)

58221.85 125652.19 24506.68 159367.36 -9208.49 193082.53


- 68 percentage of the observations have values between [58221.85  to  125652.19]

- 95 percentage of the observations have values between [24506.68  to 159367.36]

- 99.7 percentage of the observations have values between [-9208.49 to  193082.53]

In [98]:
68*200/100

136.0

In [107]:
con1 = visa_df['prevailing_wage']>val_minus_1
con2 = visa_df['prevailing_wage']<val_plus_1
len(visa_df[con1 & con2])

123

In [113]:
len(visa_df[con1 & con2])/len(visa_df)

0.615

- 68% should fall between (u - sigma) and (u + sigma) when data is normal. #(according to the above image)
- But.....
- the above data ste is not normal, so only 19.5% fall in that range. #(according to my data 19.5%)

In [109]:
con3 = visa_df['prevailing_wage']>val_minus_2
con4 = visa_df['prevailing_wage']<val_plus_2
len(visa_df[con3 & con4])

200

In [115]:
len(visa_df[con3 & con4])/len(visa_df)

1.0

In [111]:
con5 = visa_df['prevailing_wage']>val_minus_3
con6 = visa_df['prevailing_wage']<val_plus_3
len(visa_df[con5 & con6])

200

In [119]:
len(visa_df[con5 & con6])/len(visa_df)

1.0